base/system/winlogon/rpcserver.c

```diff
@@ -485,6 +507,87 @@ BaseRegQueryMultipleValues(
     return ERROR_SUCCESS;
 }
 
+INT_PTR
+CALLBACK
+ShutdownDialogProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    switch (uMsg)
+    {
+        case WM_INITDIALOG:
+        {
+            g_hShutdownDialog = hwndDlg;
+            if (g_ShutdownParams.usMessage.Length)
+                SetDlgItemTextW(hwndDlg,
+                                IDC_SHUTDOWNCOMMENT,
+                                g_ShutdownParams.usMessage.Buffer);
+            RemoveMenu(GetSystemMenu(hwndDlg, FALSE), SC_CLOSE, MF_BYCOMMAND);
+            SetWindowPos(hwndDlg, HWND_TOPMOST, 0, 0, 0, 0, SWP_NOSIZE | SWP_NOMOVE);
+            PostMessage(hwndDlg, WM_TIMER, 0, 0);
+            SetTimer(hwndDlg, IDT_SYSSHUTDOWN, 1000, NULL);
+        break;
+        }
+        case WM_CLOSE:
+        {
+            g_bShuttingDown = FALSE;
+            KillTimer(hwndDlg, IDT_SYSSHUTDOWN);
+            if (g_ShutdownParams.usMessage.Buffer)
+            {
+                HeapFree(GetProcessHeap(), 0, g_ShutdownParams.usMessage.Buffer);
```

Comment:
A `RtlInitEmptyUnicodeString(&(g_hShutdownParams.usMessage), NULL, 0);` just after this HeapFree() may be needed to turn the string empty but valid again.